build: Add linter checking for accidental introduction of locale dependence

practicalswift opened the pull request:
This linter will check for code accidentally introducing locale dependencies.

Unnecessary locale dependence can cause bugs that are very tricky to isolate and fix. We should avoid using locale dependent functions if possible.

Context: https://github.com/bitcoin/bitcoin/pull/12881#issuecomment-378564722

Example output:

```
$ contrib/devtools/lint-locale-dependence.sh
The locale dependent function tolower(...) appears to be used:
src/init.cpp:    if (s[0] == '0' && std::tolower(s[1]) == 'x') {

Unnecessary locale dependence can cause bugs that are very
tricky to isolate and fix. Please avoid using locale dependent
functions if possible.

Advice not applicable in this specific case? Add an exception
by updating the ignore list in contrib/devtools/lint-locale-dependence.sh
```

**Note to reviewers:** What is the most appropriate `LOCALE_DEPENDENT_FUNCTIONS` function list? What should be added or removed?

practicalswift committed:
build: Add linter for checking accidental locale dependence

practicalswift commented:
@laanwj Do you think the `LOCALE_DEPENDENT_FUNCTIONS` list is correct? :-)

laanwj commented:
Concept ACK. I think this is a good idea.

> @laanwj Do you think the LOCALE_DEPENDENT_FUNCTIONS list is correct? :-)

It's a good start.
Please add `strftime` too, that's the one that caused recent discussion in #12973.

practicalswift commented:
@laanwj `strftime` and `strptime` added. Please re-review :-)

laanwj commented on contrib/devtools/lint-locale-dependence.sh:
```diff
@@ -0,0 +1,31 @@
+    toupper towlower towupper
+)
+REGEXP_IGNORE_EXTERNAL="src/(crypto/ctaes/|leveldb/|secp256k1/|tinyformat.h|univalue/)"
+REGEXP_IGNORE_KNOWN_VIOLATIONS="^(src/base58.cpp:.*isspace|src/qt/rpcconsole.cpp:.*isdigit|src/rest.cpp:.*strtol|src/torcontrol.cpp:.*strtol|src/uint256.cpp:.*isspace|src/uint256.cpp:.*tolower|src/util.cpp:.*tolower|src/utilmoneystr.cpp:.*isdigit|src/utilmoneystr.cpp:.*isspace|src/utilstrencodings.cpp:.*isspace|src/utilstrencodings.cpp:.*strtoll|src/utilstrencodings.cpp:.*strtol)"
```
That's a shocking number of known violations.
I didn't know e.g. `strtol` is also locale dependent.

laanwj replied:
Is it possible to split this list over multiple lines, so that it's possible to remove an exception by removing a line?

practicalswift commented:
@laanwj Updated version with the known violations list split over multiple lines. Please re-review :-)

practicalswift commented:
The use of `strtoul(...)` and `strtoull(...)` might be problematic too?

practicalswift commented:
Updated: Added more functions to `LOCALE_DEPENDENT_FUNCTIONS`. Added two known violations for `strtoul(...)` and `strtoull(...)` in `utilstrencodings.cpp`.

Please re-review :-)

laanwj commented:
utACK 29b73b509f1aa537a835f9e75878622c3a18335f

practicalswift commented:
@theuni @sipa Your language expertise would be great in reviewing this PR if you have time – does  `LOCALE_DEPENDENT_FUNCTIONS` look correct? :-)

practicalswift commented:
Updated:
* Added more functions to `LOCALE_DEPENDENT_FUNCTIONS`.
* Added known `atoi(…)` violations in `bitcoin-tx.cpp`, `init.cpp`, `qt/rpcconsole.cpp`, `src/torcontrol.cpp`, `src/util.cpp`, `utilstrencodings.cpp` and `utilstrencodings.h`.

Please re-review :-)

laanwj commented:
I think `printf` and `fprintf(stderr, ...` are fine. They are only used to print error messages, it's not important if they are formatted slightly different according to the locale.
(not so much for `sprintf` `vsnprintf` etc which should be replaced with use of tinyformat)

laanwj commented:
> @laanwj You saw that fprintf(stderr, … is explicitly filtered out via grep -vE 'fprintf\(.*(stdout|stderr)' in the script? :-)

Oh that's good! No, hadn't noticed that yet.

> Should I keep fprintf (with the exclusion above) and remove printf? Makes sense?

Yes I think so!

practicalswift commented:
@laanwj Updated version with `printf` removed! Please re-review :-)

laanwj commented:
Thinking of it, it would make sense to add this list to the developer notes as well (next to the mention of the locale risks) , so it doesn't come out of the blue.

practicalswift committed:
docs: Mention lint-locale-dependence.sh in developer-notes.md

practicalswift commented:
@laanwj Good point! Added a reference to the script and included the list in the developer notes. Looks good? :-)

practicalswift commented:
Updated: Added some locale dependent Boost functions and the corresponding `KNOWN_VIOLATIONS` entries.

Added known violations:
* `is_digit` is used in `src/core_read.cpp`
* `to_lower` is used in `src/netbase.cpp`
* `to_upper` is used in `src/rpc/server.cpp`
* `trim` is used in `src/core_read.cpp`
* `trim` is used in `src/httprpc.cpp`
* `trim_right` is used in `src/bitcoin-tx.cpp`

Added Boost functions:

```
fold_case
is_digit
is_space
normalize
to_lower
to_title
to_upper
trim
trim_left
trim_right
```

@laanwj, is this PR getting ready for merge? :-)  When it is merged my plan is to start replacing calls to locale dependent functions with their non-locale dependent counterparts.

practicalswift commented:
@Empact Would you mind reviewing this PR? :-)

Empact commented on contrib/devtools/lint-locale-dependence.sh:
```diff
@@ -0,0 +1,229 @@
+    mbtowc       # LC_CTYPE
+    mktime
+    normalize    # boost::locale::normalize
+#   printf       # LC_NUMERIC
```
Would be helpful to add an explanatory comment as to why each of these are disabled. E.g. I don't see a printf call outside of crypto/secp256k1/tinyformat.h so it's not clear why this is disabled.

Empact commented on contrib/devtools/lint-locale-dependence.sh:
```diff
@@ -0,0 +1,229 @@
+        echo
+        EXIT_CODE=1
+    fi
+done
```
To keep the code simple, might be fine to just output the GIT_GREP_OUTPUT with some explanatory text. I'm skeptical grouping by function will be helpful.

practicalswift replied:
@Empact For performance or usability reasons?

If you mean for performance reasons, then please note that the costly operation `git grep` is run only once.

From a usability perspective I think it makes sense to group the output per function - try running it with an empty `KNOWN_VIOLATIONS` and you'll see what I mean :-)

Empact replied:
I'm commenting from a maintainability perspective. In the general case KNOWN_VIOLATIONS will be populated so the list will be relatively short and only introduced by they who open the PR, in which case the grouping isn't very beneficial IMO. Given that perspective, if not much usability benefit is gained, then if the code can be simplified by removing it I think it's worthwhile to do so.

practicalswift replied:
@Empact Do you have any specific output formatting in mind? Please give an example on how it would look :-)

laanwj commented:
utACK f947ef02f9c8eee5957bb02e12cee4edd636c9fa

cnavigato commented:
On Jun 5, 2018 11:15 AM, "Wladimir J. van der Laan" <
notifications@github.com> wrote:

utACK f947ef0
<https://github.com/bitcoin/bitcoin/commit/f947ef02f9c8eee5957bb02e12cee4edd636c9fa>

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bitcoin/bitcoin/pull/13041#issuecomment-394808844>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAfpiNwEhc18PlReCb8tP5iuXoUiYLn6ks5t5sqtgaJpZM4TdXns>
.

ken2812221 commented:
Need to move to test/lint

practicalswift commented:
@ken2812221 Thanks! Moved!

@laanwj @cnavigato Please re-review :-)

ken2812221 reviewed (approved):
utACK 698cfd081144845f6246171b8a2a0cfa8daaecdb
